### PR TITLE
Extended -r option to allow uploading recent without monitoring

### PR DESCRIPTION
**Pull request by martinemde**
Uploads all recent games that are missing from Stat.ink and then exits, allowing the script to be run repeatedly to gather and upload any recent missing battles.

Enables ENV variables for easier deployment without committing sensitive keys.

Fixes for python3 and pipenv to manage dependencies.

(Note that I'm not usually a python developer. The repetition in my code bugs my like crazy but I'm too slow at working with python to quickly refactor.)

Thank you for this script. I'm using it now in a Heroku scheduler to run every hour with these modifications.

### CODE REVIEW

File: splatnet2statink.py [hunk 1]
```diff
@@ -273,37 +328,18 @@ def monitor_battles(s_flag, t_flag, r_flag, secs, debug):
...
-	# if r_flag, check if there are any battles in splatnet that aren't on stat.ink
-	if r_flag:
-		print "Checking if there are previously-unuploaded battles..."
-		printed = False
-		url  = 'https://stat.ink/api/v2/user-battle?only=splatnet_number&count=50'
-		auth = {'Authorization': 'Bearer ' + API_KEY}
-		resp = requests.get(url, headers=auth)
-		statink_battles = json.loads(resp.text) # 50 recent battles on stat.ink
-
-	for result in results[::-1]:
```
Comment (martinemde) [inline splatnet2statink.py]: In python, does this mean "iterate through these in reverse"?

File: splatnet2statink.py [hunk 2]
```diff
@@ -239,12 +240,66 @@ def main():
...
+
+	# if r_flag, check if there are any battles in splatnet that aren't on stat.ink
+	print("Checking if there are previously-unuploaded battles...")
+	printed = False
+	url  = 'https://stat.ink/api/v2/user-battle?only=splatnet_number&count=50'
+	auth = {'Authorization': 'Bearer ' + API_KEY}
+	resp = requests.get(url, headers=auth)
+	statink_battles = json.loads(resp.text) # 50 recent battles on stat.ink
+
+	for i, result in reversed(list(enumerate(results))): # reversed chrono order
```
Comment (martinemde) [inline splatnet2statink.py]: You used this elsewhere and I copied it here to get the `i == 0` behavior for the most recent battle.

File: splatnet2statink.py (hunk 1 above)
Comment (clovervidia) [reply splatnet2statink.py]: Yes, the `-1` means to go through the array in reverse order.

**Comment by clovervidia** [comment]
Thanks for your contributions. I have a few questions about your proposed changes.

>Enables ENV variables for easier deployment without committing sensitive keys.

I understand this will allow the user to store those variables as environment variables rather than store them in the config.txt file. This might make it easier for users who are familiar with working with those variables. Why do you say this makes for easier deployments without committing sensitive keys?

Have you tested this under Python 2.7? I'm not too well-versed in the differences between the two versions aside from the parentheses when using `print`, `input` vs `raw_input`, and some of the module names, but I do most of my side projects in 3.6, and I know all of our users run the script under 2.7 because that's how it was originally written. If it can work with both versions without any difference in functionality, that would be ideal.

Being able to upload recent battles without using monitor mode seems like a pretty useful function, although to my knowledge, most users run the script in monitoring mode, so they probably wouldn't miss uploading any battles, but it looks like a nice thing to have.

I'm glad you find the script useful. When you say "run every hour", are you saying that you run the script to check for recent battles every hour?

File: splatnet2statink.py (hunk 2 above)
Comment (frozenpandaman) [reply splatnet2statink.py]: `enumerate()` provides both a counter for numbers (`for i in int i < 100…`) and things (`for string in str_list…`) . think of the two types of for loops in java and bundling them into one. technically it returns an iterator that gives { (0, thing[0]), (1, thing[1]), … }

Comment (frozenpandaman) [reply splatnet2statink.py]: `enumerate()` provides both a counter for numbers (`for int i < 100…`) and things (`for string in str_list…`) . think of the two types of for loops in java and bundling them into one. afaik technically it returns an iterator that gives `{(0, thing[0]), (1, thing[1]), …}`

**Comment by frozenpandaman** [comment]
Hi @martinemde.

>Fixes for python3

Are you running the script using Python 3, or could you explain those changes? From my understanding, Python 2.7 isn't perfectly backwards compatible with Python 3. You need to run `2to3` (or a similar tool like `futurize` or `modernize`) to fix a bunch of commands and then manually go through and make more updates/changes to get it to run. This script is meant to be run using Python 2.7, and it's set up for our users to be run that way, although feel free to make modifications to your local installation…

>manage dependencies

Please see past discussion of this (i.e. having a separate `requirements.txt` or pip setup file):
* https://github.com/frozenpandaman/splatnet2statink/issues/23#issuecomment-352272577
* https://github.com/frozenpandaman/splatnet2statink/pull/11#issuecomment-327913296

>Enables ENV variables for easier deployment without committing sensitive keys.

I think we want to shy away from using environment variables. Please see discussion of it here: https://github.com/frozenpandaman/splatnet2statink/pull/4#issuecomment-325178198

Sensitive keys will never be committed (?) as long as you're not trying to upload `config.txt` (hence its inclusion in `.gitignore` and this file being generated – it's local only).

>Uploads all recent games that are missing from Stat.ink and then exits

With the script set up as it is, I don't see the point of retroactively uploading some `n` number of forgotten battles (a static, fixed operation) at the same time/in the same command as uploading the previous `m` battles (similarly, a fixed, "one time" operation). Remember **most of our users are launching the script manually** on their local machines. In that case, this is the only scenario I could see it playing out:

1. Play 10 battles. Upload 10 recent battles.
2. Play 5 more battles.
3. Play 8 more battles. Upload 8 recent battles.
4. Oh no! I forgot to upload the battles from step two.

If users find themselves in this scenario, the solution, for me, comes at step three: next time, do a better job at manually checking how many battles you need to upload (should have uploaded 13, not 8). By the time you get to step four you can solve it in any number of ways (uploading the past 13, and keyboard interrupting the program after the un-uploaded 5; manually deleting the 8 battles from stat.ink and re-doing the 13 all in one go; etc.) I don't think it's up to us to solve problems at this step.

…or just use monitoring mode which solves these sorts of problems. Which relates to:

>Heroku scheduler to run every hour

Is there a reason you want to run the script as a scheduled task (cron job) and exit, rather than deploying it as a daemon (continuously run in background)? We could maybe accommodate this, but we'd have to go about it in a different way – by adding a launch option (e.g. `-R`, to contrast with `-M`) that **uploads all non-uploaded battles and exits** (and if you've played more than 50, oops). As it's currently set up, if you're not using `-M`/monitoring mode, the script will wait for some user input, which doesn't seem ideal here…

File: splatnet2statink.py (hunk 1 above)
Comment (frozenpandaman) [reply splatnet2statink.py]: - `[::-1]` makes a reversed copy of a list (not really necessary here, i guess…)
- `reversed()` returns an iterator and walks through a list in reverse order.
- and just for good measure `list.reverse()` is in-place list reversal.

Comment (martinemde) [reply splatnet2statink.py]: Thanks! I’m most comfortable in ruby so python isn’t far off, but syntax like this isn’t as easy to search for.

Comment (frozenpandaman) [reply splatnet2statink.py]: i 100% love ruby but i don't know ruby

MINASWAN

**Comment by martinemde** [comment]
@clovervidia The ENV change is so I can run this automatically in a free Heroku scheduler that boots up hourly to import any battles. Best practice is to not commit your API keys and set them in env following 12factor app rules. I think it’s completely transparent for users, so it could be undocumented like the file behavior you have.

@frozenpandaman

* Python 3 and Pipfile - I had to make these changes for Heroku to work. They run 3.6 with pipenv. Fine with me if they are local for me only. My goal however would be to make this a one click deploy to a Heroku scheduler that can run it for free for those who want it.

* On got point about uploading a fixed number of battles, I see it as very different. Personally I would use the monitor if I were running this with persistent state, but I can’t. Stateless monitor is essentially what I’ve written here. I think people would be relieved to not have to carefully remember how many battles to upload when we have all the information necessary to do it for them. Is there a reason we want them to choose how many to upload? Would you purposely upload 8 when there are 13 unrecorded battles? Is there a good reason to cause an issue that we can solve transparently?

Onto the`-R` suggestion. The script as submitted here does not wait for user input with only the -r option. It uploads anything Nintendo is willing to give tha is not already in stat.ink. This is what I thought `-r` did at first, which spawned this fork and PR ;) I figured out that `-M -r` does sort of what I wanted but kept running. The reason for avoiding the daemon is that it requires an always on computer. This process oriented approach handles your sleep wait loop by simply not running and using a scheduler to boot.

The updated -r here would be good for average people wanting to do what I did at first: upload all 50 battles, play a few, run it again to sync again, repeat. However, the best part is that those of us that want to can boot it in this sync mode regularly with a cron or scheduler and avoid the hassle of a long running process or even having a machine to run it on.





I’d be happy to split these commits up too. This is just everything I needed to get it running on Heroku.

**Comment by martinemde** [comment]
@frozenpandaman one more thought (and I’m typing on my phone which github makes editing the above comment difficult for no good reason).

I think essentially this script only does one thing: upload battles not already uploaded (possibly with intentional exceptions for battles you don’t want). Let’s call this function sync.

Monitor call sync every interval.
-r as it stands would call sync once.
Specifying n battles does sync of most recent with limit n.

I think no user ever wants to upload the same battle twice, so that negates the use case of specifying n=5 when only 3 are unsynced. Additionally, we only support “n most recent” so there is no easy use case for anything but a limit.

I’d suggest not thinking of them as modes. We have 3 functions really:

1. Look for missing battles in your history? (yes/no) `-r`
2. How many battles previous? (n default=50)
3. Stay running? `-M`

**Comment by frozenpandaman** [comment]
@martinemde,

>The ENV change is so I can run this automatically in a free Heroku scheduler that boots up hourly to import any battles.

>My goal however would be to make this a one click deploy to a Heroku scheduler that can run it for free for those who want it.

It's cool to maintain something like that in a fork, but we're not going to assume people are deploying to Heroku and make a change that instead complicates things for the majority of our users and/or intended userbase. If multiple other users express interest in something like this in the future we can revisit this and perhaps come up with a solution that makes that sort of setup easier by default.

>Would you purposely upload 8 when there are 13 unrecorded battles?

I often play private battles with friends where I'm just fooling around, sharing gear, etc. I think giving users some agency in this regard is good no matter their reasons, rather than restricting them. And I don't think people would have to "carefully remember" since they can just check the app versus their profile on stat.ink if they really need that level of control (otherwise, I assume they'll use monitoring mode, per its current intended use). You've already partially addressed this in your second comment, though.

>The reason for avoiding the daemon is that it requires an always on computer. 

Web server/VPS, or if you want to keep it local, Raspberry Pi, etc.? I'm not exactly sure what you meant by "persistent state" but, either way, I do understand how this isn't viable for everyone.

I agree it would be nice to be able to add this sort of feature one way or other, i.e. a way to run with `-r` alone, aka running with `-M -r` and stopping after. What about just allowing users to type -1 (or something) at the prompt in order to auto-upload all non-uploaded battles? And then we could have a `skip_prompt` [custom key](https://github.com/frozenpandaman/splatnet2statink/wiki/custom-keys) that would allow users to easily run it (bypassing the I/O; always running with -1 as the "entered" value) as a scheduled task/cron job (essentially what you're doing via Heroku if I understand correctly). That way, the desired "upload all 50 battles, play a few, run it again to sync again, repeat" workflow/usage could be attained.

**Comment by martinemde** [comment]
@frozenpandaman I understand where you're coming from. I don't mind maintaining a fork. If you'd like to see the heroku deploy button, I added it to [my repository master](https://github.com/martinemde/splatnet2statink). You can click that without incurring any cost and see what you think of the setup. It removes a lot of the friction that already exists in running this, though creating a free Heroku account still adds some.

As for `-1` that seems to me more like a magic number than a good user interface. What significance does the -1 have? I think we can make it very simple by allowing `-r` alone to have that behavior. When you use `-r` with the `-M` option, the incremental change the behavior of the script is to do exactly the single action I'm aiming for. Therefore, if you run it without -M, it makes sense that it would do *only* the -r behavior. We could avoid the prompting as well by supporting a -n option. If it's set, it will not prompt in any mode and will respect that limit.

I want the skip private battles option to be supported too, to partially solve your use case. Then you could purposely choose to upload some private battles. This draws my attention to the need for an offset. i.e. upload the 7th battle previously in my history. This could allow you to upload private battles manually that were missed by the script running in cron.

Overall, I'd like to help you improve this script, so please take any of my input purely in the pursuit of explaining my use case and helping users that might not have the ability or the desire to run this script on a local machine. Speaking of, what would you consider this to be licensed under? MIT license? Any chance of clarifying that so that I know how to properly attribute if I do maintain a fork?

**Comment by martinemde** [comment]
BTW, I suggest that we do not merge my PR as it stands. I agree that adding pip and upgrading to python3 are not necessary to add this behavior, and could hinder existing users.

**Comment by martinemde** [comment]
Ok, scratch the above comment. I've re-pushed to include only the extended `-r` option. We can discuss and implement the ENV or anything else separately, since it's not necessary for the primary behavior that I wanted to enable in this script.

**Comment by frozenpandaman** [comment]
@martinemde,

This looks pretty good to me. I'd support merging this unless @clovervidia has any further concerns. Hopefully I'll have a chance to check out Heroku stuff soon since it would be cool to have an experience bordering on "click here to use this" available to users.

Note that currently, running with `-M` and `-r` does a double check for unuploaded battles when exiting. Can we fix this? (i.e. remove the second check – everything after "Bye!")

```
Checking if there are previously-unuploaded battles...
No previously-unuploaded battles found.
Waiting for new battles... (checking every 5 minutes)
Press Ctrl+C to exit. 299 ^C
Checking to see if there are unuploaded battles before exiting...
No remaining battles found.
0 wins and 0 losses this session.
Bye!
Checking if there are previously-unuploaded battles...
No previously-unuploaded battles found.
```

We'll want to bump the version number and take into account 1b33707 as well. (Or, I could just do this at the time I merge the PR.)

>what would you consider this to be licensed under?

https://github.com/frozenpandaman/splatnet2statink#license 😁

**Comment by clovervidia** [comment]
It looks good to me as well. I'm not too well versed in Heroku, so I'm still a bit confused as to why environment variables are better for storing the sensitive keys as compared to the config file, but I think we've reached a point where we can merge once everything's tested and confirmed to be working.

**Comment by martinemde** [comment]
@frozenpandaman oh funny, I missed the license at the bottom. Thanks!

@clovervidia Heroku does not maintain a persistent filesystem. Servers are booted and destroyed as needed. To push the config.txt, I'd need to commit it to my repository. Here is the [12 Factor spec regarding configuration](https://12factor.net/config). 12 factor is [a set of guidelines](https://12factor.net/) aimed at making cloud deployment highly repeatable and scalable.

Besides security, ENV config is very well supported by Heroku, which allows me to specify the configuration for booting this application and which variables need to be defined. Here's the part of the boot interface on Heroku that asks for the keys you need before you boot.

create_new_app___heroku

It also means that anyone can deploy this code without even downloading it, and they could update to the new version by hitting "deploy" in heroku console.

**Comment by martinemde** [comment]
@frozenpandaman Is it me, or has it always done that?

https://github.com/frozenpandaman/splatnet2statink/blob/1b33707cee95fa55a0944342811e361206129ee5/splatnet2statink.py#L331-L356

There's even a comment lamenting the copied code 😛. The first run is here: 

https://github.com/frozenpandaman/splatnet2statink/blob/1b33707cee95fa55a0944342811e361206129ee5/splatnet2statink.py#L282-L297

Seems like a single function that checks for missing battles would be a good refactor.

**Comment by martinemde** [comment]
Fixed! `if` to `elif`. Glad you caught that.

**Comment by frozenpandaman** [comment]
https://github.com/martinemde/splatnet2statink/blob/extended-r-option/splatnet2statink.py#L333
```
if r_flag:
	battles = upload_recent(s_flag, t_flag, debug)
else:
	battles = [] # 50 recent battles on splatnet
```

I don't think this last line is what we want.

Because, if `battles` is an empty list and never populated, it will think we haven't uploaded any of the last 50 and, on exit, reupload all of the previous 50 starting from the last one…

I'm working on a fix now though.

**Comment by frozenpandaman** [comment]
https://github.com/martinemde/splatnet2statink/blob/feb57a595ca4eaf0eadbe84009e2c84cf69c163a/splatnet2statink.py#L330-L333
```
if r_flag:
	battles = upload_recent(s_flag, t_flag, debug)
else:
	battles = [] # 50 recent battles on splatnet
```

I don't think this last line is what we want.

Because, if `battles` is an empty list and never populated, it will think we haven't uploaded any of the last 50 and, on exit, reupload all of the previous 50 starting from the last one…

I'm working on a fix now though.

**Comment by frozenpandaman** [comment]
I'm going to roll back to this code (albeit moving it to a separate function, `upload_recent()` which I'm renaming to `populate_battles()`):

https://github.com/frozenpandaman/splatnet2statink/blob/1b33707cee95fa55a0944342811e361206129ee5/splatnet2statink.py#L280-L300

That *whole* block can't depend on `r_flag` – the for loop needs to run either way, to populate the `battles` array.

**Comment by frozenpandaman** [comment]
Merged, with the few fixes as well. Let me know if everything looks good.

**Comment by martinemde** [comment]
Thanks for figuring out the fixes. I updated my 2 copies running in heroku to your newest master version and they are working.

Glad we could make this feature happen!

**Comment by frozenpandaman** [comment]
Hopefully will have some time to check out Heroku deployment over the next few weeks, as well.

**Comment by frozenpandaman** [comment]
Something's wrong with `-r`. Not sure if it's on stat.ink's end or our fault, but something to look into…

```
>py -2 splatnet2statink.py -M -r
splatnet2statink v0.2.20
Checking if there are previously-unuploaded battles...
Previously-unuploaded battles detected. Uploading now...
Battle uploaded to https://stat.ink/@into_the_trash/spl2/299362
Battle uploaded to https://stat.ink/@into_the_trash/spl2/299363
...
```
etc. But these battles have already been uploaded.

**Comment by martinemde** [comment]
I read over the code again, and tried `-M -r` (which I expected to do nothing) and it all looks good. I'll keep watching for it, and running `-M -r` to make sure it continues to do nothing.

**Comment by frozenpandaman** [comment]
Discussion continues at #30

**Pull request merged**